### DISCUSSION TITLE
Possible Improvements to Asset Names & IDs

### DISCUSSION THREAD

**Discussion by Jpja** [CounterpartyXCP/CIPs#87] [CIPs]
All assets have a unique ID. The ID is what's inside Counterparty message data.

For named asset a mathematical two-way function links the name to the ID and vice versa.

Unfortunately, due to a technicality, the function breaks if the name begins with A.

**IMPROVEMENT 1: Reserve an ID space for names that begin with A.**
The following function shows how IDs above 1e19 can be applied to A-names.
```
 function assetid_b27(asset_name, radix = 16) {
        const k = BigInt(1e19);
        let b27_digits = ' ABCDEFGHIJKLMNOPQRSTUVWXYZ'; 
        let name_array = asset_name.split("");
        let n_bigint = BigInt(0);
        for (i = 0; i < name_array.length; i++) { 
          n_bigint *= BigInt(27);
          n_bigint += BigInt(b27_digits.indexOf(name_array[i]));
        }
        n_bigint += k;
        return n_bigint.toString(radix);
      }
```
The keys here are that base 27 is used here instead of b26 and the constant 1e19.

**IMPROVEMENT 2:  Allow 1-3 letter names**
Use the same function as above for all 1-3 letter names, even those that don't begin with A.

**IMPROVEMENT 3: Deterministic ID for subassets**
Currently the wallet decides the ID for for subassets. Light wallets need to trust APIs.
If instead the ID would be the 19-digit hash (decimal) of the subasset name (longname), wallets would no longer rely on an API to from subasset to ID (in the other direction they would still need an API but not trust it. They could verify that the ID would match).

How the number would be assigned to MYTOKEN.SUBASSET:
```
let sha256 = sha256('MYTOKEN.SUBASSET'); //'e20aa891b6fd48f4b715f7268dd51b8210dce361e794f8aebaeb301bd0ca043b'
let e19 = BigInt(1e19);
let e18 = BigInt(1e18);
let asset_id = BigInt('0x'+sha256) % (e19 - e18) + e18;;
```

New numerical (non-subasset) assets would need to be 18 digits.

The new namespace would roughly be:
```
BTC           0
XCP           1
Unused:       [2, 
Named         [BAAA, 95,428,956,661,682,177]
Numerical :   [1e17, 1e18-1]
Longnames :   [1e18, 1e19-1]
3-letter :    [1e19+1, 1e19+18278]
A-names       [1e19, ~1.01e19]
Unused:       [~1.01e19, ~1.84e19]
```

Since longnames, 3-letter and a-names will occupy a number-space currently used by numerical assets, light wallets will still need to check if an asset was registered before the upgrade, then apply the now logic. This is not very complicated, nor are the tables more than a few MB.

I already use such logic for my [Op_return builder](https://github.com/Jpja/Electrum-Counterparty/tree/master/db).

**Comment by jotapea** [comment]
Some comments about IMPROVEMENTs 1 and 2:

- I don't believe in *just* opening the namespace with A named assets or <=3 character named assets. These limitations are what make the protocol special:
  - It keeps application logic simple.
  - The asset name squatting is balanced because the asset name becomes more a unique (creative) ID than an actual word.

- BUT I do believe if 1, 2, ~~or 3~~ character named assets will be added, then these should be DIFFERENT from the current types of assets.
  - By this logic, then I would not add other 3 character assets (unless is another XCP-like burned BTC token).

About IMPROVEMENT 3, I could see some value in that. But by having to maintain the OLD logic then unfortunately I don't see much benefit of only applying this to new ones... but there is definitely something interesting here that can be applied to *new ideas*.

**Reply by Jpja** [reply]
With #3 the benefit is that light wallets will no longer need to rely/trust an API for the Longname -> ID link. After the upgrade, a light wallet can keep a table of all old subassets/IDs. This table is immutable and will never change. The wallet knows that 
1) if subasset in table, then use ID from table
2) if subasset NOT in table, then ID is hash

I already use such a table for the op_return builder but the problem now is that it always lags behind. If user enters a subasset not in the table, the wallet needs to assume that this is a new subasset and trust some API to get its ID. This leads to blind signing and is a security issue.

#1 and #2 complicates the logic, sure, but not more than some pretty simple functions. Not worth it?

**Reply by jotapea** [reply]
About 1 and 2 adding complexity I mean at the application layer. Currently it is very easy to identify that 3 letter assets are the protocol assets BTC and XCP. Names starting with A are numeric assets (which can be longname). Changing the rules to now allow any asset start with A or be less than 4 characters complicates the handling of assets for all applications... BUT if these are not the same kind of assets, then I do agree with you that the new added complexity can be worth it.

In relation to 3, I could agree to it... but I really don't think including the asset longname (which I've seen some are very long) in every asset issuance is actually a good thing. What you are proposing would require this, right?
